### PR TITLE
avrdude: always build documentation on Linux

### PR DESCRIPTION
**Pull request by nagy**
###### Description of changes

Make texlive a build-time dependency to always build the docs.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by bjornfor — CHANGES_REQUESTED** [review]
I feel the title `avrdude: repair documentation` is dishonest. Looks like it should be `avrdude: always build documentation`? And since docSupport was disabled by default due to being "a big dependency" (as commented in the Nix file), I think it requires some justification to change the default.

**Comment by nagy** [comment]
> I feel the title avrdude: repair documentation is dishonest.

The build of the documentation is currently broken because the attribute `texLive` does not exist ( anymore ?). You can try it out for yourself: 

```
$ nix-build -E "with import <nixpkgs> {}; avrdude.override { docSupport=true; }"
error: assertion '(docSupport -> (((texLive != null) && (texinfo != null)) && (texi2html != null)))' failed

       at /nix/store/p2w4q9vzlr8cqkqx5siim70jk5gq4bld-source/pkgs/development/embedded/avrdude/default.nix:6:1:

            5|
            6| assert docSupport -> texLive != null && texinfo != null && texi2html != null;
             | ^
            7|
```

> And since docSupport was disabled by default due to being "a big dependency" (as commented in the Nix file), I think it requires some justification to change the default.

The way it is now, apart from being broken, is that texlive is in `buildInputs` , making it a runtime dependency. I have placed texlive into `nativeBuildInputs`, making it a buildtime dependency. I would argue, that this increase in build-time dependency would justify, that hydra is building the pdf for all users, but of course I can hide it behind a flag again.

**Comment by bjornfor** [comment]
I'm not opposed to always building the docs, I just felt the commit message wasn't addressing the change precisely. I think in terms of big dependencies, runtime dependency bloat is the bigger concern.

Maybe reword the commit message and leave the rest of the patch as is?

**Review by bjornfor — APPROVED** [review]

**Comment by wegank** [comment]
Breaks avrdude on darwin...

File: pkgs/development/embedded/avrdude/default.nix [hunk 1]
```diff
@@ -16,14 +14,20 @@ stdenv.mkDerivation rec {
     sha256 = "sha256-pGjOefWnf11kG/zFGwYGet1OjAhKsULNGgh6vqvIQ7c=";
   };
 
-  nativeBuildInputs = [ cmake bison flex ];
+  nativeBuildInputs = [
+    cmake
+    bison
+    flex
+    util-linux
```
Comment (wegank) [inline pkgs/development/embedded/avrdude/default.nix:21]: Instead of using `util-linux` everywhere, the better option is to pick the binaries from `unixtools`.

**Comment by nagy** [comment]
> Breaks avrdude on darwin...

Is it better now? If not, can you please post logs?

**Comment by wegank** [comment]
```
[ 94%] Generating avrdude.info
Undefined subroutine &Locale::Messages::dgettext called at /nix/store/n80nqbxmqnx0k3x0kx2zj87s4xy1ia1q-texi2html-5.0/bin/texi2html line 29628.
make[2]: *** [src/doc/CMakeFiles/html.dir/build.make:78: src/doc/avrdude-html/avrdude.html] Error 2
make[1]: *** [CMakeFiles/Makefile2:343: src/doc/CMakeFiles/html.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 94%] Built target info
[ 96%] Built target dvi
[ 98%] Built target pdf
[ 98%] Built target ps
make: *** [Makefile:136: all] Error 2
```

**Comment by nagy** [comment]
I cannot fix that. I have disabled documentation building and its native dependencies on darwin for now. Hopefully, it is better now.

File: pkgs/development/embedded/avrdude/default.nix [hunk 2]
```diff
@@ -1,9 +1,7 @@
-{ lib, stdenv, fetchFromGitHub, cmake, bison, flex, libusb-compat-0_1, libelf, libftdi1, readline
-# docSupport is a big dependency, disabled by default
-, docSupport ? false, texLive ? null, texinfo ? null, texi2html ? null
-}:
-
-assert docSupport -> texLive != null && texinfo != null && texi2html != null;
+{ lib, stdenv, fetchFromGitHub, cmake, bison, flex, libusb-compat-0_1, libelf
+, libftdi1, readline
+# for documentation
```
Comment (wegank) [inline pkgs/development/embedded/avrdude/default.nix:3]: Probably add an `enableDocs ? !stdenv.isDarwin` here?

**Comment by nagy** [comment]
So with the problems on darwin, I have now reverted most of this PR and now it is just about repairing and enabling the docs on linux by default.

File: pkgs/development/embedded/avrdude/default.nix [hunk 3]
```diff
@@ -16,10 +14,14 @@ stdenv.mkDerivation rec {
     sha256 = "sha256-pGjOefWnf11kG/zFGwYGet1OjAhKsULNGgh6vqvIQ7c=";
   };
 
-  nativeBuildInputs = [ cmake bison flex ];
+  nativeBuildInputs = [ cmake bison flex ] ++ lib.optionals enableDocs [
```
Comment (wegank) [inline pkgs/development/embedded/avrdude/default.nix:17]:
```suggestion
  nativeBuildInputs = [ cmake bison flex ] ++ lib.optionals docSupport [
```

Comment (nagy) [reply pkgs/development/embedded/avrdude/default.nix]: Sorry, it's getting late here.

**Comment by sigprof** [comment]
> The way it is now, apart from being broken, is that texlive is in `buildInputs` , making it a runtime dependency. I have placed texlive into `nativeBuildInputs`, making it a buildtime dependency.

Apparently the change somehow still made texlive a runtime dependency, blowing up the closure size of avrdude in comparison to the version in `nixos-23.05`:
```
❯ nix path-info --size --closure-size --human-readable --store https://cache.nixos.org /nix/store/s9wq2kd1r2qvxjkzli0nm0cys9ip7qqv-avrdude-7.1 /nix/store/jj4j4nrda65gd2b160zcx25pqm660z3c-avrdude-7.1
/nix/store/jj4j4nrda65gd2b160zcx25pqm660z3c-avrdude-7.1	   3.0M	 102.9M
/nix/store/s9wq2kd1r2qvxjkzli0nm0cys9ip7qqv-avrdude-7.1	   4.6M	   1.1G
```
This might be caused by the presence of the following thing is `share/doc/avrdude/avrdude.ps`:
```ps
%DVIPSCommandLine:
%+ /nix/store/s318zscx6grakbs8jsmx894cijspkg7x-texlive-combined-medium-2022-final/bin/dvips
%+ -o avrdude.ps avrdude.dvi
```
Maybe just drop the PostScript file, leaving just the PDF, DVI and HTML formats?

**Comment by nagy** [comment]
Good catch, thank you.

> Maybe just drop the PostScript file, leaving just the PDF, DVI and HTML formats?

Sounds good to me.
